sdk/eventhub/Azure.Messaging.EventHubs/src/Primitives/CheckpointStore.cs

```diff
@@ -78,12 +79,38 @@ public abstract class CheckpointStore
         /// <param name="cancellationToken">A <see cref="CancellationToken" /> instance to signal a request to cancel the operation.</param>
         ///
-        public abstract Task UpdateCheckpointAsync(string fullyQualifiedNamespace,
+        [System.ComponentModel.EditorBrowsableAttribute(System.ComponentModel.EditorBrowsableState.Never)]
+        public virtual Task UpdateCheckpointAsync(string fullyQualifiedNamespace,
                                                    string eventHubName,
                                                    string consumerGroup,
                                                    string partitionId,
                                                    long offset,
                                                    long? sequenceNumber,
-                                                   CancellationToken cancellationToken);
+                                                   CancellationToken cancellationToken) => throw new NotImplementedException();
+
+        /// <summary>
+        ///   Creates or updates a checkpoint for a specific partition, identifying a position in the partition's event stream
+        ///   that an event processor should begin reading from.
+        /// </summary>
+        ///
+        /// <param name="fullyQualifiedNamespace">The fully qualified Event Hubs namespace the ownership are associated with.  This is likely to be similar to <c>{yournamespace}.servicebus.windows.net</c>.</param>
+        /// <param name="eventHubName">The name of the specific Event Hub the ownership are associated with, relative to the Event Hubs namespace that contains it.</param>
+        /// <param name="consumerGroup">The name of the consumer group the checkpoint is associated with.</param>
+        /// <param name="partitionId">The identifier of the partition the checkpoint is for.</param>
+        /// <param name="offset">The offset to associate with the checkpoint, indicating that a processor should begin reading form the next event in the stream.</param>
+        /// <param name="sequenceNumber">An optional sequence number to associate with the checkpoint, intended as informational metadata.  The <paramref name="offset" /> will be used for positioning when events are read.</param>
+        /// <param name="replicationGroupEpoch">The replication group epoch associated with this checkpoint. Used in conjunction with the sequence number if using a geo replication enabled Event Hubs namespace.</param>
+        /// <param name="processorAuthorIdentifier">The unique identifier of the processor that authored this checkpoint.</param>
+        /// <param name="cancellationToken">A <see cref="CancellationToken" /> instance to signal a request to cancel the operation.</param>
+        ///
+        public virtual Task UpdateCheckpointAsync(string fullyQualifiedNamespace,
```

Comment:
To future-proof this, we should consider passing some kind of checkpoint structure that takes this information.   That would save us from having to do these gymnastics again if we want to add to the checkpoint.

We have `EventProcessorCheckpoint`, though that has only a StartingPosition.  I wonder if we added some constructor overloads there it would work?    Something like:

```csharp
public class EventProcessorCheckpoint
{
    // Still need an empty one
    public EventProcessorCheckpoint();

    // All properties, including the starting position
    public EventProcessorCheckpoint(...);

    // All properties besides the starting position, plus 
    // an event that we use to create the position.
    public EventProcessorCheckpoint(..., EventData sourceEvent);

    // All properties except the starting position
    // plus the offset, sequence number, and replication segment
    public EventProcessorCheckpoint(..., long offset, long? sequenceNumber, string replicationSegment)
}
```

That does feel a bit much to me.  The bigger problem is that we would need to add properties to it so that the starting position parts could be read.   They're all internal at the moment.  Not sure if that would be "make them public on the starting position" or "add properties on the checkpoint that could read from the starting position internals.

Maybe we create a new `struct` type - though I don't have a good name for one and fear that it may be confusing.

What do you think?

Reply:
Not sure that I follow - but looking at `EventPosition, I realize that we have a potential issue there.   It (intentionally) represents a single position in the stream.   That can be either an offset or sequence number - but it cannot be both.

Reply:
Ah sorry I meant the set of overloads on the EventProcessorCheckpoint, let me edit

Reply:
Ok I realized I only really got half of my thoughts out there, I updated.

I think the only options to create the EventPosition right now is either FromOffset, FromSequenceNumber, or FromEnqueuedTime. I think we would just need to add a FromReplicationSegment or something?

Reply:
I think we'd just add an overload of `FromSequenceNumber` that accepts the segment.  There's no way to position a reader using just the segment - it only makes sense in the context of a sequence number.   We could get away with an optional parameter - but since there's a lot of context that we should provide around "use this if you've got a GeoDR enabled...." it probably makes more sense to push it to a dedicated overload.

Reply:
So maybe something like this?
```csharp
public class EventProcessorCheckpoint
{
    public EventProcessorCheckpoint();

    public EventProcessorCheckpoint(string fullyQualifiedNamespace, string eventHubName, string consumerGroup, string partitionId, EventPosition startingPosition, string clientIdentifier);
}
```
```csharp
public struct EventPosition : IEquatable<EventPosition>
{
    public static EventPosition FromSequenceNumber(long sequenceNumber, string replicationSegment, bool isInclusive = true) 
    {
        ...
    }
}
```

```csharp
public abstract class CheckpointStore
{
    public virtual Task UpdateCheckpointAsync(EventProcessorCheckpoint checkpoint, CancellationToken cancellationToken) 
    {}
}
```

Usage:

```csharp
var eventPosition = EventPosition.FromSequenceNumber(0, "0", true);
var checkpoint = new EventProcessorCheckpoint("name", "name", "group", "partition",  eventPosition, "Id");
await someCheckpointStoreSomewhere.UpdateCheckpointAsync(checkpoint);
```